### DISCUSSION TITLE
TypeError on node_modules lib results in failing Story

### DISCUSSION THREAD

**Discussion by awacode21** [histoire-dev/histoire#313] [Q&A] [answered]
Hi guys,

i need your help. I have a component which includes an iframe. The component is using the iframe resizer library: https://github.com/davidjbradshaw/iframe-resizer

The vue component when used in my project is running perfectly fine without any errors. But when i try to add a story to histoire for my component the story is always failing and not showing up in story tree in histoire because of the following error:

```
Error while collecting story /Users/awacode/code/customers/enbw/EKP-Frontend/histoire/components/base/BaseIFrame.story.vue:         21:35:40
TypeError: window.parent.postMessage is not a function
    at chkLateLoaded (/Users/awacode/code/customers/enbw/EKP-Frontend/node_modules/iframe-resizer/js/iframeResizer.contentWindow.js:1294:21)
    at /Users/awacode/code/customers/enbw/EKP-Frontend/node_modules/iframe-resizer/js/iframeResizer.contentWindow.js:1300:3
    at Object.<anonymous> (/Users/awacode/code/customers/enbw/EKP-Frontend/node_modules/iframe-resizer/js/iframeResizer.contentWindow.js:1303:3)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/awacode/code/customers/enbw/EKP-Frontend/node_modules/iframe-resizer/js/index.js:5:38)
```

I don't understand why, and i don't understand how to fix this.
I already excluded everything related to node_modules from tsconfig.
i also installed types for the resizer lib.
Also the file it is complaining about ("iframeResizer.contentWindow.js") is not imported by me. It is the one the iframe content will provide.

So how can i prevent histoire from type checking this stuff??

My tsconfig:
```
{
    "compilerOptions": {
        "target": "esnext",
        "useDefineForClassFields": true,
        "module": "esnext",
        "moduleResolution": "node",
        "strict": true,
        "strictNullChecks": true,
        "jsx": "preserve",
        "sourceMap": true,
        "resolveJsonModule": true,
        "esModuleInterop": true,
        "lib": ["esnext", "dom"],
        "types": ["vitest/globals"],
        "typeRoots": ["node_modules/@types", "src/typings"],
        "baseUrl": ".",
        "paths": {
            "/@/*": ["src/*"]
        }
    },
    "exclude": ["node_modules", "./server.ts", "./histoire/**/*.vue"],
    "include": [
        "./*.js",
        "./src/**/*.ts",
        "./src/**/*.d.ts",
        "./src/**/*.tsx",
        "./src/**/*.vue",
        "./vite.config.ts",
        "./styleDictionary/**/*.js",
        "./src/mocks/*.ts",
        "./src/_helpers/**/*.ts",
        "./histoire/**/*.ts"
    ]
}

```

**Comment by Akryum** [comment]
```TypeError: window.parent.postMessage is not a function```

This is not a type checking error as it's completely unrelated to TypeScript. It's the Node runtime complaining that it can't call `window.parent.postMessage` (likely because it's undefined).

**Reply by awacode21** [reply]
strange thing is, i don't include the script which would call that line. I am the parent. But this would get executed by an iframe if this iframe would include the child script, which it does not at the moment.
Also when running my app, this error does not appear, it only appears in context of histoire. I scratch my head about this thing.

**Reply by awacode21** [reply]
but thanks for your hint, will search in that direction.

**Reply by Akryum** [reply]
Collecting the stories involves executing the story files in an NodeJS environment (server-side).

**Reply by awacode21** [reply]
ah, well that is the hint i need, now it's getting clearer for me.  For sure it is undefined there! I will try some adjustments on my component and see if i am able to solve that. i will let this thread know when i found the solution. Thank you very much @Akryum much appreciated!

**Reply by awacode21** [reply]
ok i figured out which code it relates to. I have a custom directive for initializing the iframeResizer:

```
const vIframeResize = {
    mounted(el: HTMLIFrameElement, { value = {} }: DirectiveBinding) {
        if (typeof window != 'undefined') {
            el.addEventListener('load', () => iframeResizer(value, el));
        }
    },
    unmounted(el: { iFrameResizer: { removeListeners: () => void } }): void {
        el.iFrameResizer?.removeListeners();
    },
};
```
i added check around the relevant code that it should only be executed when window is not undefined, unfortunately the error still appears. When i comment out the code happening within the if, there is no error. So somehow my if condition does not do the trick.

@Akryum do you have an idea why this still gets executed on server side even if i check if window is available?

**Reply by awacode21** [reply]
@Akryum my app is also using SSR, i don't get the error there. Also the code should only be executed onMounted. and only when window is available. i don't understand why it is still executed in histoire on server side. how does a check needs to look like for histoire that it only runs this on client side

**Reply by Akryum** [reply]
`window` is available via happy-dom, you need to use [isCollecting()](https://histoire.dev/reference/client.html#iscollecting)

**Reply by awacode21** [reply]
I updated histoire and @histoire/plugin-vue to latest version to be able to use isCollecting().
Now i get the error not only for the story it relates to, i get the error now for every single story i have. So histoire is now completely broken for me.
The editor also complains that there is no exported member "isCollecting" in histoire/client, but if i check histoire/client it is in there.
My adjusted code:

```
import { iframeResizer } from 'iframe-resizer';
import { isCollecting } from 'histoire/client';

// some other stuff...

const vIframeResize = {
    mounted(el: HTMLIFrameElement, { value = {} }: DirectiveBinding) {
        el.addEventListener('load', () => {
            if (!isCollecting()) {
                iframeResizer(value, el);
            }
        });
    },
    unmounted(el: { iFrameResizer: { removeListeners: () => void } }): void {
        el.iFrameResizer?.removeListeners();
    },
};
```
<img width="1576" alt="Bildschirmfoto 2022-10-06 um 10 11 58" src="https://user-images.githubusercontent.com/30939926/194258450-a1625afb-8cac-44e4-9b82-414f0d7d9282.png">

**Reply by Akryum** [reply]
At this point it would be nice to have a runnable reproduction of the problem.

**Reply by awacode21** [reply]
@Akryum sorry for the delay, we had a death in the family and i needed some days getting back to work. Here is the requested  reproduction of the problem:
https://stackblitz.com/edit/histoire-vue3-starter-hqkoni?file=src%2FBaseIFrame.story.vue,src%2FBaseIFrame.vue

I forked it from the one from histoire docs and added my example.

**Reply by awacode21** [reply]
when commenting out 

`iframeResizer(value, el);`

there is no issue. but when not commented out, no matter what condition i use arround, this is massively failing for every single story in histoire, also for those which don't use the iframe.

**Reply by awacode21** [reply]
@Akryum seems that not only i am affected, @kristofgilicze seems to have the same issue with another lib. Sounds like a bug? should i create one? is it possible to merge this discussion to be a bug? did you have time to have a look?

**Comment by kristofgilicze** [comment]
I occured similar issue, builds just fine without histoire.

![image](https://user-images.githubusercontent.com/16563329/195789638-7a086424-627a-4db1-85c2-6cd267ed2653.png)

**Reply by awacode21** [reply]
ok, when it is not only me i guess it is some kind of a bug

**Comment by awacode21** [comment, marked as answer]
So using "isCollecting" to workaround the problem simply was not working at all. But i found another way to workaround the problem for me.
Instead of having the library import at the top of the file i am now doing a dynamic import within the mounted of the custom directive right before i want to use that library. That prevents histoire from crashing. But still i think there is some kind of a bug in histoire.

@kristofgilicze maybe this hint/workaround can help you also with your problem.
```
const vIframeResize = {
    mounted(el: HTMLIFrameElement, { value = {} }: DirectiveBinding) {
        el.addEventListener('load', async () => {
            // need to import here to not crash histoire
            const iframeResizer = (await import('iframe-resizer')).default
                .iframeResizer;
            iframeResizer(value, el);
        });
    },
    unmounted(el: { iFrameResizer: { removeListeners: () => void } }): void {
        el.iFrameResizer?.removeListeners();
    },
};
```
